File: samples/ocr_e2e/ocr_e2e.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include "ocr_e2e.hpp"

#include <chrono>

#include "common/cmdline.hpp"
#include "common/file_system.hpp"
#include "common/utils.hpp"
#include "opencv2/opencv.hpp"
using time_point = std::chrono::time_point<std::chrono::high_resolution_clock>;

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<std::string>(
      "det_model", '\0', "text detection model prefix of the model suite files",
      false, "/opt/vastai/vaststreamx/data/models/det_model_vacc_fp16/mod");
  args.add<std::string>("det_config", '\0',
                        "text detection vdsp preprocess parameter file", false,
                        "../data/configs/dbnet_rgbplanar.json");
  args.add<std::string>(
      "cls_model", '\0',
      "text classification model prefix of the model suite files", false,
      "/opt/vastai/vaststreamx/data/models/cls_model_vacc_fp16/mod");
  args.add<std::string>("cls_config", '\0',
                        "text classification vdsp preprocess parameter file",
                        false, "../data/configs/crnn_rgbplanar.json");
  args.add<std::string>(
      "rec_model", '\0',
      "text recognition model prefix of the model suite files", false,
      "/opt/vastai/vaststreamx/data/models/rec_model_vacc_fp16/mod");
  args.add<std::string>("rec_config", '\0',
                        "text recognition vdsp preprocess parameter file",
                        false, "../data/configs/crnn_rgbplanar.json");

  args.add<std::string>("det_box_type", '\0', "text detection box type", false,
                        "quad");
  args.add<std::string>(
      "det_elf_file", '\0', "text detection elf file", false,
      "/opt/vastai/vaststreamx/data/elf/find_contours_ext_op");
  args.add<std::string>("cls_labels", '\0', "text classification label list",
                        false, "[0, 180]");
  args.add<float>("cls_thresh", '\0', "text classification thresh", false, 0.9);
  args.add<std::string>("rec_label_file", '\0', "text recognition label file",
                        false, "../data/labels/ppocr_keys_v1.txt");
  args.add<float>("rec_drop_score", '\0',
                  "text recogniztion drop score threshold", false, 0.5);
  args.add<bool>("use_angle_cls", '\0', "use text classification", false, true);
  args.add<uint32_t>("batch_size", '\0', "batch size of the model", false, 1);
  args.add<std::string>("device_ids", '\0', "device id to run", false, "[0]");
  args.add<std::string>("hw_config", '\0', "hw-config file of the model suite",
                        false);
  args.add<std::string>("input_file", '\0', "input image", false,
                        "../data/images/word_336.png");
  args.add<std::string>("output_file", '\0', "output image file", false, "");
  args.add<std::string>("dataset_filelist", '\0', "input dataset filelist",
                        false, "");
  args.add<std::string>("dataset_root", '\0', "input dataset root", false, "");
  args.add<std::string>("dataset_output_file", '\0', "dataset output file",
                        false, "dataset_output.txt");
  args.parse_check(argc, argv);
  return args;
}

void InferenceThread(std::shared_ptr<vsx::OCR_e2e> model, cmdline::parser& args,
                     std::mutex& merge_mutex, std::vector<int64_t>& costs,
                     uint32_t device_id, float& throughput) {
  vsx::SetDevice(device_id);

  // test one image
  if (args.get<std::string>("dataset_filelist").empty()) {
    auto cv_image = cv::imread(args.get<std::string>("input_file"));
    CHECK(!cv_image.empty())
        << "Failed to read image:" << args.get<std::string>("input_file")
        << std::endl;
    auto result = model->Process(cv_image);
    if (result.empty()) {
      std::cout << "No object detected in image:"
                << args.get<std::string>("input_file") << std::endl;
    } else {
      std::cout << "Thread " << device_id << " get "
                << args.get<std::string>("input_file") << " result:\n";
      for (auto& item : result) {
        auto coor = std::get<0>(item);
        auto score = std::get<1>(item);
        auto str = std::get<2>(item);
        std::cout << "bbox:[ [" << static_cast<int>(coor[0]) << " "
                  << static_cast<int>(coor[1]) << "] ["
                  << static_cast<int>(coor[2]) << " "
                  << static_cast<int>(coor[3]) << "] ["
                  << static_cast<int>(coor[4]) << " "
                  << static_cast<int>(coor[5]) << "] [ "
                  << static_cast<int>(coor[6]) << " "
                  << static_cast<int>(coor[7]) << "] ], score: " << score
                  << ", string: " << str << std::endl;
      }
      if (args.get<std::string>("output_file") != "") {
        for (auto& item : result) {
          auto coor = std::get<0>(item);
          auto str = std::get<2>(item);
          cv::line(cv_image, cv::Point2f(coor[0], coor[1]),
                   cv::Point2f(coor[2], coor[3]), cv::Scalar(0, 0, 255));
          cv::line(cv_image, cv::Point2f(coor[2], coor[3]),
                   cv::Point2f(coor[4], coor[5]), cv::Scalar(0, 0, 255));
          cv::line(cv_image, cv::Point2f(coor[4], coor[5]),
                   cv::Point2f(coor[6], coor[7]), cv::Scalar(0, 0, 255));
          cv::line(cv_image, cv::Point2f(coor[0], coor[1]),
                   cv::Point2f(coor[6], coor[7]), cv::Scalar(0, 0, 255));
        }
        fs::path output_path = args.get<std::string>("output_file");
        auto dir = output_path.parent_path().string();
        if (dir.empty()) dir = ".";
        auto filename = output_path.filename().string();
        auto save_file =
            dir + "/thread_" + std::to_string(device_id) + "_" + filename;
        std::cout << "Save file to: " << save_file << std::endl;
        cv::imwrite(save_file, cv_image);
      }
    }
    return;
  }

  // test dataset
  std::vector<std::string> filelist =
      vsx::ReadFileList(args.get<std::string>("dataset_filelist"));
  auto dataset_root = args.get<std::string>("dataset_root");
  fs::path output_path = args.get<std::string>("dataset_output_file");
  auto dir = output_path.parent_path().string();
  auto filename = output_path.filename().string();
  auto save_file =
      dir + "/thread_" + std::to_string(device_id) + "_" + filename;
  std::ofstream outfile(save_file, std::ios::out);
  CHECK(outfile.is_open()) << "Failed to open " << save_file;
  std::vector<time_point> ticks;
  std::vector<time_point> tocks;
  for (size_t s = 0; s < filelist.size(); s++) {
    auto fullname = filelist[s];
    if (!dataset_root.empty()) fullname = dataset_root + "/" + fullname;
    std::cout << "Thread: " << device_id << "," << fullname << std::endl;
    auto cv_image = cv::imread(fullname);
    ticks.push_back(std::chrono::high_resolution_clock::now());
    auto result = model->Process(cv_image);
    tocks.push_back(std::chrono::high_resolution_clock::now());

    for (auto& item : result) {
      auto coor = std::get<0>(item);
      auto score = std::get<1>(item);
      auto str = std::get<2>(item);
      outfile << "bbox:[ [" << static_cast<int>(coor[0]) << " "
              << static_cast<int>(coor[1]) << "] [" << static_cast<int>(coor[2])
              << " " << static_cast<int>(coor[3]) << "] ["
              << static_cast<int>(coor[4]) << " " << static_cast<int>(coor[5])
              << "] [ " << static_cast<int>(coor[6]) << " "
              << static_cast<int>(coor[7]) << "] ], score: " << score
              << ", string: " << str << std::endl;
    }
  }
  outfile.close();

  if (ticks.size() != tocks.size()) {
    std::cout << "Error! ticks.size() != tocks.szie(). ticks.size() = "
              << ticks.size() << ", tocks.size() = " << tocks.size()
              << std::endl;
  }

  int64_t cost_sum = 0;
  merge_mutex.lock();
  for (size_t i = 0; i < ticks.size(); i++) {
    auto cost = std::chrono::duration_cast<std::chrono::milliseconds>(tocks[i] -
                                                                      ticks[i])
                    .count();
    cost_sum += cost;
    costs.push_back(cost);
  }
  throughput += ticks.size() * 1000.0f / cost_sum;
  merge_mutex.unlock();
}

int main(int argc, char** argv) {
  auto args = ArgumentParser(argc, argv);
  auto cls_labels = vsx::ParseVecUint(args.get<std::string>("cls_labels"));
  auto device_ids = vsx::ParseVecUint(args.get<std::string>("device_ids"));

  std::vector<std::shared_ptr<vsx::OCR_e2e>> models;
  models.reserve(device_ids.size());
  for (auto device_id : device_ids) {
    auto model = std::make_shared<vsx::OCR_e2e>(
        args.get<std::string>("det_model"), args.get<std::string>("det_config"),
        args.get<std::string>("det_box_type"),
        args.get<std::string>("det_elf_file"),
        args.get<std::string>("cls_model"), args.get<std::string>("cls_config"),
        cls_labels, args.get<float>("cls_thresh"),
        args.get<std::string>("rec_model"), args.get<std::string>("rec_config"),
        args.get<std::string>("rec_label_file"),
        args.get<float>("rec_drop_score"), args.get<bool>("use_angle_cls"),
        args.get<uint32_t>("batch_size"), device_id,
        args.get<std::string>("hw_config"));
    models.push_back(model);
  }

  std::mutex merge_mutex;
  std::vector<int64_t> costs;
  std::vector<std::thread> threads;
  float throughput = 0;
  threads.reserve(device_ids.size());

  for (size_t i = 0; i < device_ids.size(); i++) {
    threads.emplace_back(InferenceThread, models[i], std::ref(args),
                         std::ref(merge_mutex), std::ref(costs), device_ids[i],
                         std::ref(throughput));
  }

  for (auto& thread : threads) {
    thread.join();
  }

  // test one image
  if (args.get<std::string>("dataset_filelist").empty()) {
    return 0;
  }
  // test dataset
  int64_t cost_sum = 0;
  for (auto cost : costs) cost_sum += cost;
  float avg_cost = cost_sum * 1.0 / costs.size();
  cost_sum /= device_ids.size();

  std::cout << "Image count: " << costs.size() << ", total cost: " << cost_sum
            << " ms, throughput: " << throughput
            << " fps. Average latency: " << avg_cost << " ms. " << std::endl;
  return 0;
}

File: samples/ocr_e2e/ocr_e2e.py
#
# Copyright (C) 2025 Vastai-tech Company.
# All rights reserved.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
#
import os
import sys

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../..")
sys.path.append(common_path)

import common.utils as utils
import numpy as np
import cv2
import argparse
from common.text_det import TextDetector
from common.text_cls import TextClassifier
from common.text_rec import TextRecognizer
import copy
import vaststreamx as vsx
from easydict import EasyDict as edict
import time
import threading
import ast

attr = vsx.AttrKey


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--det_model",
        default="/opt/vastai/vaststreamx/data/models/dbnet_resnet50_vd-int8-kl_divergence-1_3_736_1280-vacc/mod",
        help="text detection model prefix of the model suite files",
    )
    parser.add_argument(
        "--det_vdsp_params",
        default="./data/configs/dbnet_rgbplanar.json",
        help="text detection vdsp preprocess parameter file",
    )
    parser.add_argument(
        "--det_box_type",
        default="quad",
        help="det box type, poly or quad",
    )
    parser.add_argument(
        "--det_elf_file",
        default="/opt/vastai/vaststreamx/data/elf/find_contours_ext_op",
        help="input file",
    )
    parser.add_argument(
        "--cls_model",
        default="/opt/vastai/vaststreamx/data/models/resnet34_vd-int8-max-1_3_32_100-vacc/mod",
        help="text detection model prefix of the model suite files",
    )
    parser.add_argument(
        "--cls_vdsp_params",
        default="./data/configs/crnn_rgbplanar.json",
        help="text detection vdsp preprocess parameter file",
    )
    parser.add_argument(
        "--cls_label_list",
        type=list,
        default=["0", "180"],
        help="text classification label list",
    )
    parser.add_argument(
        "--cls_thresh", type=float, default=0.9, help="text classification thresh"
    )
    parser.add_argument(
        "--rec_model",
        default="/opt/vastai/vaststreamx/data/models/resnet34_vd-int8-max-1_3_32_100-vacc/mod",
        help="text detection model prefix of the model suite files",
    )
    parser.add_argument(
        "--rec_vdsp_params",
        default="./data/configs/crnn_rgbplanar.json",
        help="text detection vdsp preprocess parameter file",
    )
    parser.add_argument(
        "--rec_label_file",
        default="../data/labels/key_37.txt",
        help="text recognizition label file",
    )
    parser.add_argument(
        "--rec_drop_score",
        type=float,
        default=0.5,
        help="text recogniztion drop score threshold",
    )
    parser.add_argument(
        "--use_angle_cls", type=bool, default=True, help="whether use angle classifier"
    )
    parser.add_argument(
        "--hw_config",
        default="",
        help="hw-config file of the model suite",
    )
    parser.add_argument(
        "--device_ids",
        default="[0]",
        type=str,
        help="device ids to run",
    )
    parser.add_argument(
        "--input_file",
        default="../../../data/images/detect.jpg",
        help="input file",
    )
    parser.add_argument(
        "--output_file",
        default="./ocr_result.jpg",
        help="output file",
    )
    parser.add_argument(
        "--dataset_filelist",
        default="",
        help="dataset filelist",
    )
    parser.add_argument(
        "--dataset_root",
        default="",
        help="input dataset root",
    )
    parser.add_argument(
        "--dataset_output_file",
        default="",
        help="dataset output file",
    )
    args = parser.parse_args()
    return args


class OCR_e2e:
    def __init__(
        self,
        det_model,
        det_config,
        det_box_type,
        det_elf_file,
        cls_model,
        cls_config,
        cls_label_list,
        cls_thresh,
        rec_model,
        rec_config,
        rec_label_file,
        rec_drop_score,
        use_angle_cls,
        batch_size=1,
        device_id=0,
        hw_config="",
    ):
        self.text_det = TextDetector(
            det_model,
            det_config,
            batch_size,
            device_id,
            hw_config,
            elf_file=det_elf_file,
        )
        self.text_cls = TextClassifier(
            cls_model, cls_config, cls_label_list, batch_size, device_id, hw_config
        )
        self.text_rec = TextRecognizer(
            rec_model, rec_config, rec_label_file, batch_size, device_id, hw_config
        )
        self.det_box_type = det_box_type
        self.use_angle_cls = use_angle_cls
        self.cls_thresh = cls_thresh
        self.rec_drop_score = rec_drop_score
        self.device_id = device_id
        self.image_format = self.text_det.get_fusion_op_iimage_format()

    def process(self, image, cv_image):
        [dt_boxes, dt_scores] = self.text_det.process(image)
        if dt_boxes is None or dt_boxes.size == 0:
            return None
        img_crop_list = []
        vacc_img_crop_list = []

        for bno in range(len(dt_boxes)):
            tmp_box = copy.deepcopy(dt_boxes[bno])
            if args.det_box_type == "quad":
                img_crop = self.get_rotate_crop_image(cv_image, tmp_box)
            else:
                img_crop = self.get_minarea_rect_crop(cv_image, tmp_box)
            img_crop_list.append(img_crop)
            vacc_img_crop = utils.cv_bgr888_to_vsximage(
                img_crop, self.image_format, self.device_id
            )
            vacc_img_crop_list.append(vacc_img_crop)
        if self.use_angle_cls and self.text_cls:
            cls_result = self.text_cls.process(vacc_img_crop_list)
            for rno in range(len(cls_result)):
                label, score = cls_result[rno]
                if "180" in label and score > self.cls_thresh:
                    img_crop_list[rno] = cv2.rotate(img_crop_list[rno], 1)
                    vacc_img_crop_list[rno] = utils.cv_bgr888_to_vsximage(
                        img_crop_list[rno], self.image_format, self.device_id
                    )

        rec_res = self.text_rec.process(vacc_img_crop_list)

        filter_boxes, filter_rec_res = [], []
        for box, rec_result in zip(dt_boxes, rec_res):
            text, score = rec_result[0]
            if score >= self.rec_drop_score:
                filter_boxes.append(box)
                filter_rec_res.append(rec_result)
        return filter_boxes, filter_rec_res

    def get_rotate_crop_image(self, img, points):
        """
        img_height, img_width = img.shape[0:2]
        left = int(np.min(points[:, 0]))
        right = int(np.max(points[:, 0]))
        top = int(np.min(points[:, 1]))
        bottom = int(np.max(points[:, 1]))
        img_crop = img[top:bottom, left:right, :].copy()
        points[:, 0] = points[:, 0] - left
        points[:, 1] = points[:, 1] - top
        """
        assert len(points) == 4, "shape of points must be 4*2"
        points = np.array(points, dtype=np.float32)
        img_crop_width = int(
            max(
                np.linalg.norm(points[0] - points[1]),
                np.linalg.norm(points[2] - points[3]),
            )
        )
        img_crop_height = int(
            max(
                np.linalg.norm(points[0] - points[3]),
                np.linalg.norm(points[1] - points[2]),
            )
        )
        pts_std = np.float32(
            [
                [0, 0],
                [img_crop_width, 0],
                [img_crop_width, img_crop_height],
                [0, img_crop_height],
            ]
        )
        M = cv2.getPerspectiveTransform(points, pts_std)
        dst_img = cv2.warpPerspective(
            img,
            M,
            (img_crop_width, img_crop_height),
            borderMode=cv2.BORDER_REPLICATE,
            flags=cv2.INTER_CUBIC,
        )
        dst_img_height, dst_img_width = dst_img.shape[0:2]
        if dst_img_height * 1.0 / dst_img_width >= 1.5:
            dst_img = np.rot90(dst_img)
        return dst_img

    def get_minarea_rect_crop(self, img, points):
        bounding_box = cv2.minAreaRect(np.array(points).astype(np.int32))
        points = sorted(list(cv2.boxPoints(bounding_box)), key=lambda x: x[0])

        index_a, index_b, index_c, index_d = 0, 1, 2, 3
        if points[1][1] > points[0][1]:
            index_a = 0
            index_d = 1
        else:
            index_a = 1
            index_d = 0
        if points[3][1] > points[2][1]:
            index_b = 2
            index_c = 3
        else:
            index_b = 3
            index_c = 2

        box = [points[index_a], points[index_b], points[index_c], points[index_d]]
        crop_img = self.get_rotate_crop_image(img, np.array(box))
        return crop_img

    def get_fusion_op_iimage_format(self):
        return self.text_det.get_fusion_op_iimage_format()


def inference_thread(model, args, context, thread_index):
    vsx.set_device(model.device_id)
    image_format = model.get_fusion_op_iimage_format()

    if args.dataset_filelist == "":
        cv_image = cv2.imread(args.input_file)
        assert cv_image is not None, f"Failed to read input file: {args.input_file}"
        vsx_image = utils.cv_bgr888_to_vsximage(cv_image, image_format, model.device_id)
        ocr_res = model.process(vsx_image, cv_image)
        if ocr_res is None:
            print("Do not detect any text")
        else:
            boxes, rec_res = ocr_res
            for box, rec_result in zip(boxes, rec_res):
                out_str = "["
                for point in box:
                    out_str += f"[{point[0]},{point[1]}], "
                out_str = out_str[:-2] + "], "
                print(out_str, rec_result)
            if args.output_file != "":
                for box in boxes:
                    for i in range(len(box)):
                        t = (i + 1) % len(box)
                        pt1 = (box[i][0], box[i][1])
                        pt2 = (box[t][0], box[t][1])
                        cv2.line(cv_image, pt1, pt2, color=(0, 0, 255))
                dir, basename = os.path.split(args.output_file)
                save_file = os.path.join(dir, f"thread_{thread_index}_{basename}")
                cv2.imwrite(save_file, cv_image)
                print("save file ", save_file)

    else:
        costs = []
        filelist = []
        with open(args.dataset_filelist, "rt") as f:
            filelist = f.readlines()
        dir, basename = os.path.split(args.dataset_output_file)
        save_file = os.path.join(dir, f"thread_{thread_index}_{basename}")
        with open(save_file, "wt") as outfile:
            for filename in filelist:
                fullname = os.path.join(args.dataset_root, filename.replace("\n", ""))
                print("fullname:", fullname)
                cv_image = cv2.imread(fullname)
                assert cv_image is not None, f"Read image failed:{filename}"
                vsx_image = utils.cv_bgr888_to_vsximage(
                    cv_image, image_format, model.device_id
                )
                start = time.time()
                ocr_res = model.process(vsx_image, cv_image)
                costs.append(time.time() - start)
                result_str = ""
                if ocr_res is None:
                    result_str = "Do not detect any text"
                else:
                    boxes, rec_res = ocr_res
                    for box, rec_result in zip(boxes, rec_res):
                        res_str = "["
                        for point in box:
                            res_str += f"[{point[0]},{point[1]}], "
                        res_str = (
                            res_str[:-2]
                            + "], "
                            + rec_result[0][0]
                            + ", "
                            + str(rec_result[0][1])
                        )
                        result_str += res_str + "\n"
                basename, _ = os.path.splitext(os.path.basename(fullname))
                outfile.write(f"{basename}\n{result_str}\n")
        outfile.close()
        cost_sum = np.sum(costs)
        throughput = len(costs) / cost_sum

        context.merge_lock.acquire()
        context.costs += costs
        context.throughput += throughput
        context.merge_lock.release()


if __name__ == "__main__":
    args = argument_parser()
    device_ids = ast.literal_eval(args.device_ids)

    models = []
    for id in device_ids:
        model = OCR_e2e(
            args.det_model,
            args.det_vdsp_params,
            args.det_box_type,
            args.det_elf_file,
            args.cls_model,
            args.cls_vdsp_params,
            args.cls_label_list,
            args.cls_thresh,
            args.rec_model,
            args.rec_vdsp_params,
            args.rec_label_file,
            args.rec_drop_score,
            args.use_angle_cls,
            batch_size=1,
            device_id=id,
            hw_config=args.hw_config,
        )
        models.append(model)

    threads = []
    context = edict(merge_lock=threading.Lock(), costs=[], throughput=0)

    for mod, id in zip(models, device_ids):
        thread = threading.Thread(
            target=inference_thread, args=(mod, args, context, id)
        )
        thread.start()
        threads.append(thread)

    for thread in threads:
        thread.join()

    if args.dataset_filelist == "":
        exit(0)
    costs = context.costs
    throughput = context.throughput
    avg_cost = np.mean(costs)
    cost_sum = np.sum(costs) / len(device_ids)
    print(
        f"Image count: {len(costs)}, total cost: {cost_sum:.2f} s, throughput: {throughput:.2f} fps, average latency: {avg_cost:.3f} s"
    )
